libs/pubsub/query/query.go

```diff
@@ -82,209 +89,283 @@ const (
 	TimeLayout = time.RFC3339
 )
 
-// Conditions returns a list of conditions.
-func (q *Query) Conditions() []Condition {
-	conditions := make([]Condition, 0)
+// Conditions returns a list of conditions. It returns an error if there is any
+// error with the provided grammar in the Query.
+func (q *Query) Conditions() ([]Condition, error) {
+	var (
+		eventAttr string
+		op        Operator
+	)
 
+	conditions := make([]Condition, 0)
 	buffer, begin, end := q.parser.Buffer, 0, 0
 
-	var tag string
-	var op Operator
-
-	// tokens must be in the following order: tag ("tx.gas") -> operator ("=") -> operand ("7")
+	// tokens must be in the following order: event attribute ("tx.gas") -> operator ("=") -> operand ("7")
 	for _, token := range q.parser.Tokens() {
 		switch token.pegRule {
-
 		case rulePegText:
 			begin, end = int(token.begin), int(token.end)
+
 		case ruletag:
-			tag = buffer[begin:end]
+			eventAttr = buffer[begin:end]
+
 		case rulele:
 			op = OpLessEqual
+
 		case rulege:
 			op = OpGreaterEqual
+
 		case rulel:
 			op = OpLess
+
 		case ruleg:
 			op = OpGreater
+
 		case ruleequal:
 			op = OpEqual
+
 		case rulecontains:
 			op = OpContains
+
 		case rulevalue:
 			// strip single quotes from value (i.e. "'NewBlock'" -> "NewBlock")
 			valueWithoutSingleQuotes := buffer[begin+1 : end-1]
-			conditions = append(conditions, Condition{tag, op, valueWithoutSingleQuotes})
+			conditions = append(conditions, Condition{eventAttr, op, valueWithoutSingleQuotes})
+
 		case rulenumber:
 			number := buffer[begin:end]
 			if strings.ContainsAny(number, ".") { // if it looks like a floating-point number
 				value, err := strconv.ParseFloat(number, 64)
 				if err != nil {
-					panic(fmt.Sprintf("got %v while trying to parse %s as float64 (should never happen if the grammar is correct)",
-						err,
-						number))
+					err = fmt.Errorf(
```

Comment:
may be a good idea to just write 

```
return nil, fmt.Errorf(...
```

to avoid variable shadowing

Reply:
Then you get linter errors. If you reformat the return to:

```go
return nil, fmt.Errorf(
  "got %v while trying to parse %s as int64 (should never happen if the grammar is correct)",
  err, number,
)
```

the linting is fixed, but then the return is harder to grok. Up to you.

Reply:
let's stick with your version